File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();

        vector<int> vec(nums);

        sort(vec.begin(), vec.end());

        int idx = 0;
        unordered_map<int, int>numberIdx;

        unordered_map<int, list<int>> nayaNums;


        numberIdx[vec[0]] = idx;

        nayaNums[idx].push_back(vec[0]);


        for(int i = 1; i < n; i++) {

            if(abs(vec[i] - vec[i - 1]) > limit ) {
                idx++;
            }
            numberIdx[vec[i]] = idx;
            nayaNums[idx].push_back(vec[i]);
        }


        for(int i = 0; i < n; i++) {

            idx = numberIdx[nums[i]];

            int num = *(nayaNums[idx].begin());
            nayaNums[idx].pop_front();

            nums[i] = num;
        }

        return nums;


    }
};

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.js
/**
 * @param {number[]} nums
 * @param {number} limit
 * @return {number[]}
 */
var lexicographicallySmallestArray = function(nums, limit) {


    const copyNums = [...nums].sort((a, b) => a - b);
    const number_map = new Map()
    const Number_to_index_map = new Map();


    let indx = 0;

    Number_to_index_map.set(copyNums[0], indx)
    number_map.set(indx, [copyNums[0]]);
 


    for(let i = 1; i < nums.length; i++) {

        if(Math.abs(copyNums[i] - copyNums[i - 1]) > limit){
            indx++;
        
            number_map.set(indx, [])
        }
        number_map.get(indx).push(copyNums[i]);
        Number_to_index_map.set(copyNums[i], indx);

    }

    const i_map= new Array(indx + 1).fill(0);

    for(let i = 0; i < nums.length; i++) {

        indx =  Number_to_index_map.get(nums[i]);

        nums[i] = number_map.get(indx)[i_map[indx]++]
        

    }
    return nums;
    
};